x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -0,0 +1,740 @@
+                        + "]",
+                    e
+                );
+                urlException = e;
+            }
+        } else {
+            urlException = new Exception(
+                "URL ["
+                    + jwkSetPath
+                    + "] is not valid for setting ["
+                    + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
+                    + "]. Must use HTTPS or local file."
+            );
+        }
+        // Suppress any Path parsing exception. If needed, it will be added to a SettingsException at the end.
+        Exception pathException = null;
+        try {
+            // Use separate lines, in case any are null. That allows NPE stack trace line number to show exactly which one is null.
+            final Environment environment = realmConfig.env();
+            final Path directoryPath = environment.configFile();
+            final Path filePath = directoryPath.resolve(jwkSetPath);
+            // Check file exists, accessible, and non-empty. TODO Maybe return parsed contents here instead of later in a watcher thread.
+            final String fileContents = Files.readString(filePath, StandardCharsets.UTF_8);
+            if (Strings.hasText(fileContents) == false) {
+                throw new Exception(
+                    "Empty file ["
+                        + jwkSetPath
+                        + "] for setting ["
+                        + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
+                        + "]"
+                );
+            }
+            return new Tuple<>(null, filePath); // RETURN URL AS NULL AND PATH AS NON-NULL (i.e. URL Exception is discarded)
+        } catch (Exception e) {
+            pathException = new Exception(
+                "Error loading file ["
+                    + jwkSetPath
+                    + "] for setting ["
+                    + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
+                    + "]",
+                e
+            );
+        }
+        // Throw SettingsException with the two suppressed exceptions.
+        final SettingsException settingsException = new SettingsException(
+            "Invalid value [" + jwkSetPath + "] for setting " + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
+        );
+        settingsException.addSuppressed(urlException);
+        settingsException.addSuppressed(pathException);
+        throw settingsException;
+    }
+
+    private void ensureExpectedValueForInitialized(final boolean expectedValue) {
+        if (this.initialized != expectedValue) {
+            if (expectedValue) {
+                throw new IllegalStateException("Realm has not been initialized");
+            } else {
+                throw new IllegalStateException("Realm has already been initialized");
+            }
+        }
+    }
```

Comment:
Given there's only 1 call where the `expectedValue` is false, I would lean towards having a simple `ensureInitialized()` method, and inlining the one case where were check for  `"Realm has already been initialized"`

```suggestion
    private void ensurenInitialized() {
        if (this.initialized == false) {
            throw new IllegalStateException("Realm has not been initialized");
        }
    }
```

That said, no other realm bothers tracking an `initialized` state. They only check for double initialization because it would overwrite the value of an existing field (`delegatedRealms`) so it's clearly a bug to call it twice. But the lifecycle pretty much ensures we don't have large scale initialization errors.

Reply:
TODO

Reply:
Resolved